File: web/src/app/models/show.ts
import {LibraryFile} from '@app/models/library-file';
import {Cast, Crew, Genre} from '@app/models/movie';
import {Item} from '@app/models/item';

export interface Creator {
  id: number;
  credit_id: string;
  name: string;
  gender?: number;
  profile_path?: string;
}

export interface Network {
  id: number;
  name: string;
  logo_path?: string;
  origin_country: string;
}

export interface Season {
  id: number;
  name: string;
  overview?: string;
  poster_path?: string;
  air_date?: string;
  episode_count: number;
  season_number: number;
}

export interface ShowDetails {
  id: number;
  created_by: Creator[];
  genres: Genre[];
  networks: Network[];
  number_of_episodes: number;
  number_of_seasons: number;
  seasons: Season[];
  credits: {
    cast: Cast[];
    crew: Crew[];
  };
}

export interface Episode {
  air_date: string;
  crew: Crew[];
  episode_number: number;
  name: string;
  overview: string;
  season_number: number;
  still_path?: string;
  vote_average: number;
  vote_count: number;
}

export interface Show extends Item {
  type: 'show';
  id: number;
  name: string;
  original_name: string;
  original_language: string;
  origin_country: string[];
  first_air_date: string;
  poster?: string;
  backdrop?: string;
  overview: string;
  vote_average: number;
  files: LibraryFile[];
  details?: ShowDetails;
  episodes: Episode[];
}

File: src/main/scala/net/easyflix/model/Show.scala
package net.easyflix.model

import net.easyflix.json.JsonSupport
import net.easyflix.tmdb.TVDetails.{Creator, Network, Season}
import net.easyflix.tmdb.common.{Credits, Genre}
import spray.json.RootJsonFormat

case class Show(
    id: Int,
    name: String,
    original_name: String,
    original_language: String,
    origin_country: List[String],
    first_air_date: String,
    poster: Option[String],
    backdrop: Option[String],
    overview: String,
    vote_average: Float,
    files: Set[LibraryFile],
    details: Option[Show.Details],
    episodes: List[Episode]) {

  def withDetails(details: Show.Details): Show = copy(details = Some(details))

  def withEpisode(episode: Episode): Show = copy(episodes = episodes :+ episode)

  def withFiles(files: Set[LibraryFile]): Show =
    copy(files = this.files ++ files.filter(f => !this.files.map(_.path).contains(f.path)))

}

object Show extends JsonSupport {

  implicit val format: RootJsonFormat[Show] = jsonFormat13(Show.apply)

  case class Details(
      id: Int,
      created_by: List[Creator],
      genres: List[Genre],
      networks: List[Network],
      number_of_episodes: Int,
      number_of_seasons: Int,
      seasons: List[Season],
      credits: Credits)

  object Details {

    implicit val format: RootJsonFormat[Details] = jsonFormat8(Details.apply)

  }

}
